routing/turns_tts_text.cpp

```diff
@@ -62,20 +122,139 @@ std::string GetTtsText::GetTurnNotification(Notification const & notification) c
   if (notification.m_useThenInsteadOfDistance && notification.m_turnDir == CarDirection::None)
     return {};
 
-  std::string const dirStr = GetTextById(GetDirectionTextId(notification));
   if (dirStr.empty())
     return {};
 
-  std::string distStr;
+  std::string thenStr;
   if (notification.m_useThenInsteadOfDistance)
   {
-    distStr = GetTextById("then");
-    if (notification.m_distanceUnits > 0)
-      distStr = distStr + " " + GetTextById(GetDistanceTextId(notification));
+    // add "then" and space only if needed, for appropriate languages
+    thenStr = GetTextByIdTrimmed("then");
+    if (localeKey != "ja")
+      thenStr.push_back(' ');
+  }
+
+  std::string distStr;
+  if (notification.m_distanceUnits > 0)
+    distStr = GetTextByIdTrimmed(GetDistanceTextId(notification));
+
+  // Get a string like 245; CA 123; Highway 99; San Francisco
+  // In the future we could use the full RoadNameInfo struct to do some nice formatting.
+  std::string streetOut;
+  RouteSegment::RoadNameInfo nsi = notification.m_nextStreetInfo;  // extract non-const
+  FormatFullRoadName(nsi, streetOut);
+
+  if (!streetOut.empty())
+  {
+    // We're going to pronounce the street name.
+
+    // Replace any full-stop characters (in between sub-instructions) to make TTS flow better.
+    // Full stops are: . (Period) or 。 (East Asian) or । (Hindi)
+    RemoveLastDot(distStr);
+
+    // If the turn direction with the key +_street exists for this locale, and isn't "NULL",
+    // use it (like make_a_right_turn_street)
+    std::string dirStreetStr = GetTextByIdTrimmed(dirKey + "_street");
+    if (!dirStreetStr.empty() && dirStreetStr != "NULL")
+      dirStr = std::move(dirStreetStr);
+
+    // Normally use "onto" for "turn right onto Main St" unless it's "NULL"
+    std::string ontoStr = GetTextByIdTrimmed("onto");
```

Comment:
Will returning an empty string instead of artificial NULL simplify the code?

Reply:
@biodranik No, we've gone back and forth on this about 5 times: if an i18n string is actually empty it will be detected by various i18n validators as missing and either fail validation or encourage people to supply it which is not necessary for most languages. But with increasing i18n complexity we do want translators to be able to provide i18n'd formatter string values as appropriate. So the default for some of these strings will be the keyword `"NULL"` so as to make it explicit that we want this string to be blank for languages without another string defined. This is my workaround to hacking into our i18n system as discussed in chat..